scientific-computing-with-python/learn-inheritance-by-building-an-equation-solver/quadratic_equation_solver.py

```diff
@@ -0,0 +1,145 @@
+    def grade(self):
+        return 1
+    
+    @property
+    def type(self):
+        return 'Linear Equation'
+
+    def solve(self):
+        x = - self.coefficients[0] / self.coefficients[1]
+        self.results.append(f'x = {round(x, 3)}')
+        return [x]
+    
+    def analyze(self):
+        self.details.append(f'slope = {self.coefficients[1]:>16}\ny-intercept = {self.coefficients[0]:>10}')
+        return {'slope': self.coefficients[1], 'intercept': self.coefficients[0]}
+
+
+class QuadraticEquation(Equation):
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.delta = self.coefficients[1]**2 - 4 * self.coefficients[2] * self.coefficients[0]
+
+    @property
+    def grade(self):
+        return 2
+    
+    @property
+    def type(self):
+        return 'Quadratic Equation'
+
+    def solve(self):
+        if self.delta < 0:
+            self.results.append('No real roots')            
+            return []
+
+        x1 = (- self.coefficients[1] + (self.delta)**0.5)/(2 * self.coefficients[2])
+        x2 = (- self.coefficients[1] - (self.delta)**0.5)/(2 * self.coefficients[2])        
+        if self.delta == 0:
+            self.results.append(f'x = {round(x1, 3)}')            
+            return [x1]
+
+        self.results.extend([f'x1 = {round(x1, 3)}', f'x2 = {round(x2, 3)}'])        
+        return [x1, x2]
+
+    def analyze(self):
+        x = - self.coefficients[1] / (2 * self.coefficients[2])
+        y = self.coefficients[2] * x**2 + self.coefficients[1] * x + self.coefficients[0]
+        vertex = {'x': x, 'y': y}                    
+        if self.coefficients[2] > 0:
+            concavity = 'upwards'
+            vertex['m'] = 'min'            
+        else:
+            concavity = 'downwards'
+            vertex['m'] = 'max'
+        coord = f'{(round(x, 3), round(x, 3))}'
+        self.details.append(f'concavity = {concavity:>12}\n{vertex["m"]} = {coord:>18}')
+        return {'vertex': vertex, 'concavity': concavity}
+    
+
+class Solver:
+    def __init__(self, equation):
+        if not isinstance(equation, Equation):
+            raise TypeError("Argument must be an Equation object")
+        self.equation = equation
+
+    def _solve(self):
+        self.equation.solve()
+        self.equation.analyze()
+    
+    def print_output(self):
+        self._solve()
+        output_string = '\n{:-^24}'.format(self.equation.type) + f'\n\n{str(self.equation):^24}\n\n'        
+        output_string += '{:-^24}'.format('Solutions')
+        output_string += '\n\n'
+        for result in self.equation.results:
+            output_string += f'{result:^24}\n'
+        output_string += '\n'        
+        output_string += '{:-^24}'.format('Details')
+        output_string += '\n\n' 
+        for detail in self.equation.details:
+            output_string += detail
+        output_string += '\n'
+        print(output_string)
+        
+
+eq = QuadraticEquation(-2, -3, 2)
```

Comment:
The general formula for a quadratic equation is `ax**2 + bx + c = 0`, the coefficients should also be mapped in this sequence.

For example, I would interpret `eq = QuadraticEquation(-2, -3, 9)` as `-2x**2 - 3x + 9` instead of `9x**2 - 3x - 2`.

Currently it maps to `c+bx+ax**2`. It's not wrong, just different from the standard equation.